pkg/utils/backups/sqlite3.go

```diff
@@ -0,0 +1,101 @@
+package backups
+
+import (
+	"database/sql"
+	"path/filepath"
+	"time"
+
+	sqlite3 "github.com/mattn/go-sqlite3"
+	"github.com/rotationalio/ensign/pkg/utils/sqlite"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	SQLite3Main   = "main"
+	PagesPerStep  = 5
+	StepSleep     = 50 * time.Millisecond
+	DefaultDBName = "backup.db"
+)
+
+// SQLite3 implements a single sqlite3 backup that uses the online backup of a running
+// database mechanism to copy the db over to a second sqlite3 database at the temporary
+// directory location.
+type SQLite3 struct {
+	DB   *sqlite.Conn
+	Name string
+}
+
+var _ Backup = &SQLite3{}
+
+// Backup executes the sqlite3 backup strategy.
+func (s *SQLite3) Backup(tmpdir string) (err error) {
+	var (
+		dstDB   *sql.DB
+		dstConn *sqlite.Conn
+	)
+
+	// Open a second sqlite3 database at the backup location.
+	if dstDB, dstConn, err = s.OpenDestDB(tmpdir); err != nil {
+		return err
+	}
+
+	// Ensure the database connection is closed when the backup is complete; this will
+	// also close the underlying sqlite3 connection.
+	defer dstDB.Close()
+
+	// Create the backup manager into the destination db from the src connection.
+	// NOTE: backup.Finish() MUST be called to prevent panics.
```

Comment:
If `backup.Finish()` must be called, should we just `defer` it?

Reply:
The problem with deferring is that we can't catch the error it returns, and there are some errors that relate to db corruption that are important to catch. It does make the code a bit harder to maintain but is necessary unfortunately.